packages/python-packages/apiview-copilot/src/_retry.py

```diff
@@ -0,0 +1,46 @@
+def retry_with_backoff(
+    func,
+    max_retries=5,
+    retry_exceptions=(Exception,),
+    on_failure=None,
+    on_retry=None,
+    logger=None,
+    description="operation",
+):
+    """
+    Generic retry function with exponential backoff.
+
+    Args:
+        func: The function to retry
+        max_retries: Maximum number of retries
+        retry_exceptions: Tuple of exceptions that should trigger a retry
+        on_failure: Function to call on final failure (params: exception, attempt)
+        on_retry: Function to call on each retry (params: exception, attempt, max_retries)
+        logger: Logger object to use
+        description: Description of the operation for logging
+
+    Returns:
+        The result of the function call, or the result of on_failure if all retries fail
+    """
+    for attempt in range(max_retries):
+        try:
+            result = func()
+            return result
+        except retry_exceptions as e:
```

Comment:
Should we narrow this to specific retryable exceptions?

Reply:
Yeah, good point. Kinda pointless to retry 5 times if you send a bad request or something like that.